implants/lib/eldritch/src/file/list_impl.rs

```diff
@@ -0,0 +1,182 @@
+use anyhow::Result;
+use starlark::{values::{dict::Dict, Heap, Value}, collections::SmallMap, const_frozen_string};
+use sysinfo::{System, SystemExt, UserExt};
+use std::fs::DirEntry;
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+#[cfg(target_os = "macos")]
+use std::os::macos::fs::MetadataExt;
+#[cfg(target_os = "linux")]
+use std::os::linux::fs::MetadataExt;
+#[cfg(target_os = "windows")]
+use std::os::windows::fs::MetadataExt;
+use chrono::{Utc, DateTime, NaiveDateTime};
+use super::{File, FileType};
+
+const UNKNOWN: &str = "UNKNOWN";
+
+// https://stackoverflow.com/questions/6161776/convert-windows-filetime-to-second-in-unix-linux
+fn windows_tick_to_unix_tick(windows_tick: u64) -> i64{
+    const WINDOWS_TICK: u64 = 10000000;
+    const SEC_TO_UNIX_EPOCH: u64 = 11644473600;
+    return (windows_tick / WINDOWS_TICK - SEC_TO_UNIX_EPOCH) as i64;
+}
+
+fn create_file_from_dir_entry(dir_entry: DirEntry) -> Result<File> {
+    let mut sys = System::new();
+    sys.refresh_users_list();
+
+    let file_name = match dir_entry.file_name().into_string() {
+        Ok(local_file_name) => local_file_name,
+        Err(_) => return Err(anyhow::anyhow!("file.list: Unable to convert file name to string.")),
+    };
+
+    let mut file_type = FileType::Unknown;
+    let tmp_file_type = dir_entry.file_type().expect("file.list: Unable to determine type");
+    if tmp_file_type.is_dir() {
+        file_type = FileType::Directory;
+    } else if tmp_file_type.is_file() {
+        file_type = FileType::File;
+    } else if tmp_file_type.is_symlink() {
+        file_type = FileType::Link;
+    }
+
+    let dir_entry_metadata = dir_entry.metadata().expect("file.list: Unable to get dir_entry metadata.");
+
+    let file_size = dir_entry_metadata.len();
+
+    #[cfg(unix)]
+    let owner_username = {
+        let tmp_owner_uid = dir_entry_metadata.st_uid();
+        let tmp_sysinfo_owner_uid = sysinfo::Uid::try_from(tmp_owner_uid as usize).expect("file.list: Failed to convert u32 to UID");
+        let owner_username = sys.get_user_by_id(&tmp_sysinfo_owner_uid).expect("file.list: Failed to resolve username from UID").name().to_string();
+        owner_username
+    };
+    #[cfg(not(unix))]
+    let owner_username = {
+        UNKNOWN.to_string()
+    };
+
+    #[cfg(unix)]
+    let group_id = {
+        dir_entry_metadata.st_gid().to_string()
+    };
+    #[cfg(not(unix))]
+    let group_id = {
+        UNKNOWN // This is bad but windows file ownership is very different.
```

Comment:
sadge

Reply:
Yeah we'll have to do a bunch of windows API calls to get file perms and they're also formatted differently which is annoying from the language perspective. I'll have to poke around more and we'll try to figure out how to proceed.